table/block_based_table_reader.cc

```diff
@@ -1352,6 +1368,22 @@ Status BlockBasedTable::PutDataBlockToCache(
     assert(raw_block_contents->is_raw_block);
 #endif  // NDEBUG
 
+    // Block fetcher always use memory allocator from regular block cache.
+    // Copy the content to compressed block cache.
+    //
+    // If there's a way to know if the block to read from file is compressed
+    // or not, we can save this memory copy by using the right memory
+    // allocator in BlockFetcher.
+    if (memory_allocator != memory_allocator_compressed) {
+      size_t block_size = raw_block_contents->data.size();
+      size_t alloc_size = block_size + kBlockTrailerSize;
+      CacheAllocationPtr alloc_comp =
+          AllocateBlock(alloc_size, memory_allocator_compressed);
+      memcpy(alloc_comp.get(), raw_block_contents->allocation.get(),
+             alloc_size);
+      raw_block_contents->data = Slice(alloc_comp.get(), block_size);
+      raw_block_contents->allocation = std::move(alloc_comp);
```

Comment:
Actually, unless `may_compress` is false, we should not allocate memory for uncompressed blocks from block cache.